src/backend/distributed/executor/local_executor.c

```diff
@@ -506,3 +507,31 @@ DisableLocalExecution(void)
 					  (superuser() ? PGC_SUSET : PGC_USERSET), PGC_S_SESSION,
 					  GUC_ACTION_LOCAL, true, 0, false);
 }
+
+
+/*
+ * GenerateShardQueryStringIfMissing gets the workerJob's query and the task.
+ * The function generates task->queryString if missing.
+ *
+ * For performance reasons, the queryString is not generated for
+ * local-fast path queries. However, during the execution, the
+ * executor might decide not to execute the task locally. In that case,
+ * re-generate the queryString.
+ */
+char *
+TaskQueryString(Task *task)
+{
+	if (task->queryStringLazy != NULL)
+	{
+		return task->queryStringLazy;
+	}
+	Assert(task->query != NULL);
```

Comment:
Also, might be useful to have a comment explaining why we're doing the context switch as it might not be obvious for the readers